File: backend/app/schemas/order.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.base import MoneyField, NonNegativeMoney, PositiveMoney


class OrderRequest(BaseModel):
    market_id: str
    outcome: str = Field(...)  # Accepts "yes", "no", or outcome name for multi-outcome markets
    side: str = Field(..., pattern="^(buy|sell)$")
    order_type: str = Field(default="market", pattern="^(market|limit|fill_or_kill)$")
    amount: PositiveMoney
    price: NonNegativeMoney | None = Field(None, le=1)  # required for limit
    expires_at: datetime | None = None  # for limit orders
    post_only: bool = False  # if True, reject if would execute immediately
    client_order_id: str | None = None
    max_slippage: NonNegativeMoney | None = Field(None, le=1)  # e.g. 0.005 = 0.5%
    min_shares_out: PositiveMoney | None = None  # minimum shares to receive
    quote_id: str | None = None  # bind to a specific quote


class QuoteRequest(BaseModel):
    market_id: str
    outcome: str = Field(...)
    side: str = Field(..., pattern="^(buy|sell)$")
    amount: PositiveMoney


class QuoteResponse(BaseModel):
    quote_id: str
    market_id: str
    outcome: str
    side: str
    amount: MoneyField
    price: MoneyField
    slippage: MoneyField
    yes_price: MoneyField
    no_price: MoneyField
    expires_at: float  # unix timestamp


class OrderResponse(BaseModel):
    id: str
    market_id: str
    outcome: str
    side: str
    order_type: str
    amount: MoneyField
    price: MoneyField
    status: str
    shares_bought: MoneyField | None
    shares_sold: MoneyField | None
    fee: MoneyField | None
    created_at: datetime

    model_config = {"from_attributes": True}


class PositionResponse(BaseModel):
    id: str
    market_id: str
    market_question: str | None
    outcome: str
    shares_held: MoneyField
    average_price: MoneyField
    realized_pnl: MoneyField
    unrealized_pnl: MoneyField

    model_config = {"from_attributes": True}

File: frontend/apps/web/hooks/api/types/order.ts
// Quote response — matches backend QuoteResponse
export interface QuoteResponse {
  quote_id: string
  market_id: string
  outcome: string
  side: string
  amount: string
  price: string
  slippage: string
  yes_price: string
  no_price: string
  expires_at: number // unix timestamp float
}

// Order — matches backend OrderResponse + frontend display extras
export interface Order {
  id: string
  market_id: string
  outcome: string
  side: string
  order_type: string
  status: string
  amount: string
  price: string
  shares_bought: string | null
  shares_sold: string | null
  fee: string | null
  quote_id: string | null
  client_order_id: string | null
  created_at: string
  expires_at: string | null
  // Frontend display extras (not from backend)
  market_question?: string
  remaining_amount?: string
}

// OrdersResponse — backend includes total
export interface OrdersResponse {
  success: boolean
  data: {
    orders: Order[]
    total: number
    page: number
    page_size: number
    has_more: boolean
  }
}

// Position — matches backend PositionResponse
// market_slug kept (frontend uses for routing, not from backend)
export interface Position {
  id: string
  market_id: string
  market_slug: string
  market_question: string | null
  outcome: string
  shares_held: string
  average_price: string
  realized_pnl: string
  unrealized_pnl: string
}

// PositionsResponse — backend does NOT include total
export interface PositionsResponse {
  success: boolean
  data: {
    positions: Position[]
    page: number
    page_size: number
    has_more: boolean
  }
}
